fuzz/src/refund_deser.rs

```diff
@@ -0,0 +1,103 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+use bitcoin::secp256k1::{KeyPair, PublicKey, Secp256k1, SecretKey, self};
+use crate::utils::test_logger;
+use core::convert::Infallible;
+use lightning::chain::keysinterface::EntropySource;
+use lightning::ln::PaymentHash;
+use lightning::ln::features::BlindedHopFeatures;
+use lightning::offers::invoice::{BlindedPayInfo, UnsignedInvoice};
+use lightning::offers::parse::SemanticError;
+use lightning::offers::refund::Refund;
+use lightning::onion_message::BlindedPath;
+use lightning::util::ser::Writeable;
+
+#[inline]
+pub fn do_test<Out: test_logger::Output>(data: &[u8], _out: Out) {
+	if let Ok(bech32_encoded) = std::str::from_utf8(data) {
+		if let Ok(refund) = bech32_encoded.parse::<Refund>() {
+			let bech32_encoded = refund.to_string();
+			assert_eq!(refund, bech32_encoded.parse::<Refund>().unwrap());
+
+			let secp_ctx = Secp256k1::new();
+			let keys =
+				KeyPair::from_secret_key(&secp_ctx, &SecretKey::from_slice(&[42; 32]).unwrap());
+			let pubkey = PublicKey::from(keys);
+			let mut buffer = Vec::new();
+
+			if let Ok(invoice) = build_response(&refund, pubkey, &secp_ctx) {
+				invoice
+					.sign::<_, Infallible>(
+						|digest| Ok(secp_ctx.sign_schnorr_no_aux_rand(digest, &keys))
+					)
+					.unwrap()
+					.write(&mut buffer)
+					.unwrap();
+			}
+		}
+	}
+}
+
+struct Randomness;
+
+impl EntropySource for Randomness {
+	fn get_secure_random_bytes(&self) -> [u8; 32] { [42; 32] }
+}
+
+fn pubkey(byte: u8) -> PublicKey {
+	let secp_ctx = Secp256k1::new();
+	PublicKey::from_secret_key(&secp_ctx, &privkey(byte))
+}
+
+fn privkey(byte: u8) -> SecretKey {
+	SecretKey::from_slice(&[byte; 32]).unwrap()
+}
+
+fn build_response<'a, T: secp256k1::Signing + secp256k1::Verification>(
+	refund: &'a Refund, signing_pubkey: PublicKey, secp_ctx: &Secp256k1<T>
+) -> Result<UnsignedInvoice<'a>, SemanticError> {
+	let entropy_source = Randomness {};
+	let paths = vec![
+		BlindedPath::new(&[pubkey(43), pubkey(44), pubkey(42)], &entropy_source, secp_ctx).unwrap(),
+		BlindedPath::new(&[pubkey(45), pubkey(46), pubkey(42)], &entropy_source, secp_ctx).unwrap(),
+	];
+
+	let payinfo = vec![
+		BlindedPayInfo {
+			fee_base_msat: 1,
+			fee_proportional_millionths: 1_000,
+			cltv_expiry_delta: 42,
+			htlc_minimum_msat: 100,
+			htlc_maximum_msat: 1_000_000_000_000,
+			features: BlindedHopFeatures::empty(),
+		},
+		BlindedPayInfo {
+			fee_base_msat: 1,
+			fee_proportional_millionths: 1_000,
+			cltv_expiry_delta: 42,
+			htlc_minimum_msat: 100,
+			htlc_maximum_msat: 1_000_000_000_000,
+			features: BlindedHopFeatures::empty(),
+		},
+	];
+
+	let payment_paths = paths.into_iter().zip(payinfo.into_iter()).collect();
+	let payment_hash = PaymentHash([42; 32]);
+	refund.respond_with(payment_paths, payment_hash, signing_pubkey)?.build()
```

Comment:
> Can you provide an input that crashes the fuzzer with the `unwrap` changes?

Here's one for an offer that should cause a crash.

```rust
// Encoded offer
b"lno1pgpkvmm0psgqgqqqqqqqqqqqqqqqqqqqqqqqq93pqdd7t628sgykwj5kuc837qmlv9m9gr7sq8ap6erfgacv26nhp8zzc"
```

Let me know if you need something similar for the others as it will require a bit more work.

> Also, does your comment apply on both the `respond_with` call _and_ the `build` call? Similar for invoice_request/offer.

Just `respond_with` / `request_invoice`.

Reply:
> Would love some test cases, yes. Seeding the fuzzer with some examples should improve coverage a lot. I'd really like to see better coverage before landing this.
> 

```rust
// bech32-encoded refund
b"lnr1qqsqzqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqg2qdnx7m6jqgp7s4qsqsqqqqqqqqqqqqqqqqqqqqqqqpvzzq6muh550qsfva9fdes0ruph7ctk2s8aqq06r4jxj3msc448wzwy9s"
```

```rust
// hex-encoded invoice request
"00202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a080203e80a03666f6f162102bb58b5feca505c74edc000d8282fc556e51a1024fc8e7d7e56c6f887c5c8d5f25410040000000000000000000000000000005821035be5e9478209674a96e60f1f037f6176540fd001fa1d64694770c56a7709c42cf040aae781a58c335e148e86a9de5e8d29467bd8fe2385ac1b62dd5c1bfec4a082a4dd2ce6e244c1b3d57cbc3b5417ee72d6a0582292bb398186e3c6bfdba71dfae9"
```

> Cool, can you change the `?`s that you think aren't reachable in all the fuzzers to `unwrap`s?

Done. Also, looks like `build` can fail if the offer / refund is already expired.

```diff
$ git diff-tree -U3 48ea1ec..4e008ade
diff --git a/fuzz/src/offer_deser.rs b/fuzz/src/offer_deser.rs
index f0753100..d4802bb4 100644
--- a/fuzz/src/offer_deser.rs
+++ b/fuzz/src/offer_deser.rs
@@ -47,14 +47,14 @@ fn build_response<'a>(
 	let mut builder = offer.request_invoice(vec![42; 64], pubkey)?;
 
 	builder = match offer.amount() {
-		None => builder.amount_msats(1000)?,
+		None => builder.amount_msats(1000).unwrap(),
 		Some(Amount::Bitcoin { amount_msats }) => builder.amount_msats(amount_msats + 1)?,
 		Some(Amount::Currency { .. }) => return Err(SemanticError::UnsupportedCurrency),
 	};
 
 	builder = match offer.supported_quantity() {
-		Quantity::Bounded(n) => builder.quantity(n.get())?,
-		Quantity::Unbounded => builder.quantity(10)?,
+		Quantity::Bounded(n) => builder.quantity(n.get()).unwrap(),
+		Quantity::Unbounded => builder.quantity(10).unwrap(),
 		Quantity::One => builder,
 	};
 

```

Reply:
Here's a parseable invoice with unknown feature bits, though nothing is checking for those as of yet.

```rust
// // hex-encoded invoice
"00202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a080203e80a03666f6f162102bb58b5feca505c74edc000d8282fc556e51a1024fc8e7d7e56c6f887c5c8d5f25821035be5e9478209674a96e60f1f037f6176540fd001fa1d64694770c56a7709c42ca0fd01c402ba72a6e8ba53e8b971ad0c9823968aef4d78ce8af255ab43dff83003c902fb8d035c4e0dec7215e26833938730e5e505aa62504da85ba57106a46b5a2404fc9d8e0202bb58b5feca505c74edc000d8282fc556e51a1024fc8e7d7e56c6f887c5c8d5f2002b00000000000000000000000000000000000000000000000000000000000000000000000000000000000000028f5304e2373e56ee7d774cb89e9f1afecf0ee7e3e3757f189908f069daa36c60002c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002ba72a6e8ba53e8b971ad0c9823968aef4d78ce8af255ab43dff83003c902fb8d035c4e0dec7215e26833938730e5e505aa62504da85ba57106a46b5a2404fc9d8e0203948b53da97fdf674c0877315acbcc8761aa3b9a582b439982fbafac99f97210f002d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003d98a7def67e8370118f636620d0be08ed63344df68082b4f2f1844e7ce601483002e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a23800000001000003e8002a0000000000000064000000e8d4a51000000000000001000003e8002a0000000000000064000000e8d4a510000000a40463e2cb2aa8202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2aaa0203e8ae1004000000000000000000000000000000b02102bb58b5feca505c74edc000d8282fc556e51a1024fc8e7d7e56c6f887c5c8d5f2f040103bc801dd72a80b872aa39398aae578c585640fd460d5712d34f752d700ffbe933d03fb3efc1dfbb15d467d1e99224c5fc5fd5d0d332a585a6a331a9d948b96"
```